lib/handlers/formatting.js

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2018, Compiler Explorer Authors
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+const exec = require('../exec'),
+    _ = require('underscore'),
+    logger = require('../logger').logger;
+
+class Formatter {
+    constructor(ceProps) {
+        this.tools = {};
+        this.ceProps = ceProps;
+        const formatters = _.compact(ceProps('formatters').split(':'));
+        _.each(formatters, this.fetchToolInfo.bind(this));
+    }
+
+    fetchToolInfo(formatter) {
+        const exe = this.ceProps(`formatter.${formatter}.exe`);
+        if (!exe) {
+            logger.warn(`Formatter ${formatter} does not have a valid exe. Skipping...`);
+            return;
+        }
+        const versionArg = this.ceProps(`formatter.${formatter}.version`, '--version');
+        const versionRe = this.ceProps(`formatter.${formatter}.versionRe`, '.*');
+        this.run(exe, [versionArg], {})
+            .then(result => {
+                const match = result.stdout.match(versionRe);
+                this.tools[formatter] = {
+                    exe: exe,
+                    version: match ? match[0] : result.stdout,
+                    name: this.ceProps(`formatter.${formatter}.name`, exe),
+                    styles: this.ceProps(`formatter.${formatter}.styles`, "").split(':')
+                };
+            });
+    }
+
+    supportsStyle(tool, style) {
+        return tool.styles.includes(style);
+    }
+
+    formatHandler(req, res) {
```

Comment:
This function has two concerns. One handling error cases then the actual formatting. It would be good to split them :)